### PR TITLE
Add UDF to show worker status

### PR DESCRIPTION
**Pull request by xuebinsu**
UDF `diskquota.wait_for_worker_new_epoch()` is designed to check the status of a bgworker. 

If the UDF returns successfully, the bgworker currently works well. Otherwise, there must be some issues with the bgworker.

### CODE REVIEW

File: diskquota.c [hunk 1]
```diff
@@ -230,7 +219,7 @@ define_guc_variables(void)
 							NULL,
 							&diskquota_naptime,
 							2,
-							1,
+							0,
```
Comment (xuebinsu) [inline diskquota.c:222]:
```suggestion
							1,
```

File: tests/regress/sql/test_worker_epoch.sql [hunk 2]
```diff
@@ -0,0 +1,5 @@
+-- Test if the UDF returns successfully.
+-- NOTE: This test should be the first one since the UDF is supposed
+--       to be used in all other tests.
+
+SELECT diskquota.wait_for_worker_advance();
```
Comment (xuebinsu) [inline tests/regress/sql/test_worker_epoch.sql:5]:
```suggestion
SELECT diskquota.wait_for_worker_advance();

```

File: diskquota.c [hunk 3]
```diff
@@ -1029,3 +1015,45 @@ is_valid_dbid(Oid dbid)
...
+	LWLockRelease(diskquota_locks.worker_map_lock);
+	return epoch;
+}
+
+PG_FUNCTION_INFO_V1(show_worker_epoch);
+Datum
+show_worker_epoch(PG_FUNCTION_ARGS)
+{
+	PG_RETURN_UINT32(worker_get_epoch(MyDatabaseId));
+}
```
Comment (xuebinsu) [inline diskquota.c:1059]:
```suggestion
}

```

File: diskquota--2.0.sql [hunk 4]
```diff
@@ -198,3 +198,24 @@ WITH relation_cache AS (
...
+CREATE OR REPLACE FUNCTION diskquota.wait_for_worker_advance() RETURNS boolean STRICT
+AS $$
+DECLARE
+        begin_epoch bigint;
+        current_epoch bigint;
+BEGIN
+        begin_epoch := (SELECT diskquota.show_worker_epoch());
+        LOOP
+                current_epoch := (SELECT diskquota.show_worker_epoch());
+                IF (current_epoch - begin_epoch) > 0 THEN
```
Comment (higuoxing) [inline diskquota--2.0.sql:215]: Is it possible that `show_worker_epoch()` overflows? If so, what will happen? e.g.,

```
begin_epoch := INT32_MAX;
```

Comment (xuebinsu) [reply diskquota--2.0.sql]: Turns out that it does not matter. But we need to think about how to test this.

File: diskquota.c [hunk 5]
```diff
@@ -1029,3 +1015,45 @@ is_valid_dbid(Oid dbid)
...
+	bool found = false;
+	unsigned int epoch = 0;
+	DiskQuotaWorkerEntry * workerentry = (DiskQuotaWorkerEntry *) hash_search(
+		disk_quota_worker_map, (void *) &database_oid, HASH_FIND, &found);
+	
+	if (found)
+	{
+		epoch = workerentry->epoch;
+	}
+	LWLockRelease(diskquota_locks.worker_map_lock);
```
Comment (xuebinsu) [inline diskquota.c:1050]: Otherwise should throw an error.

File: diskquota.c [hunk 6]
```diff
@@ -1008,8 +991,11 @@ start_worker_by_dboid(Oid dbid)
 	{
 		workerentry->handle = handle;
 		workerentry->pid = pid;
+		workerentry->epoch = UINT32_MAX;
```
Comment (xuebinsu) [inline diskquota.c:994]:
```suggestion
		workerentry->epoch = 0;
```

File: diskquota--2.0.sql [hunk 7]
```diff
@@ -198,3 +198,37 @@ WITH relation_cache AS (
...
+AS $$
+DECLARE
+        current_epoch bigint;
+        new_epoch bigint;
+BEGIN
+        current_epoch := (SELECT diskquota.show_worker_epoch());
+        LOOP
+                new_epoch := (SELECT diskquota.show_worker_epoch());
+                IF (new_epoch < current_epoch) THEN
+                        new_epoch := new_epoch + (1::bigint << 32);
```
Comment (xuebinsu) [inline diskquota--2.0.sql:227]:
```suggestion
                        new_epoch := new_epoch + (1::bigint << 32);
               END IF;
```

File: diskquota--2.0.sql [hunk 8]
```diff
@@ -198,3 +198,37 @@ WITH relation_cache AS (
...
+CREATE OR REPLACE FUNCTION diskquota.wait_for_worker_new_epoch() 
+RETURNS boolean STRICT
+AS $$
+DECLARE
+        current_epoch bigint;
+        new_epoch bigint;
+BEGIN
+        current_epoch := (SELECT diskquota.show_worker_epoch());
+        LOOP
+                new_epoch := (SELECT diskquota.show_worker_epoch());
```
Comment (xuebinsu) [inline diskquota--2.0.sql:225]:
```suggestion
                new_epoch := diskquota.show_worker_epoch();
```

File: diskquota--2.0.sql [hunk 9]
```diff
@@ -198,3 +198,37 @@ WITH relation_cache AS (
...
+-- 2. If it does not terminate, there must be some issues with the bgworker. 
+--    In this case, we must ensure this UDF can be interrupted by the user. 
+CREATE OR REPLACE FUNCTION diskquota.wait_for_worker_new_epoch() 
+RETURNS boolean STRICT
+AS $$
+DECLARE
+        current_epoch bigint;
+        new_epoch bigint;
+BEGIN
+        current_epoch := (SELECT diskquota.show_worker_epoch());
```
Comment (xuebinsu) [inline diskquota--2.0.sql:223]:
```suggestion
        current_epoch := diskquota.show_worker_epoch();
```

File: quotamodel.c [hunk 10]
```diff
@@ -478,6 +478,18 @@ disk_quota_shmem_startup(void)
 	if (!found)
 		memset((void *) diskquota_hardlimit, 0, sizeof(bool));
 
+	/* use disk_quota_worker_map to manage diskquota worker processes. */
+	memset(&hash_ctl, 0, sizeof(hash_ctl));
+	hash_ctl.keysize = sizeof(Oid);
+	hash_ctl.entrysize = sizeof(DiskQuotaWorkerEntry);
+	hash_ctl.hash = oid_hash;
+
+	disk_quota_worker_map = ShmemInitHash("disk quota worker map",
```
Comment (xuebinsu) [inline quotamodel.c:487]: Forgot to `add_size(hash_estimate_size())` for `disk_quota_worker_map`.

**Comment by higuoxing** [comment]
Almost done. I'm running out of time today, will continue to work on this PR tomorrow.

**Review by higuoxing — DISMISSED** [review]

**Comment by zhrt123** [comment]
+1

**Review by higuoxing — APPROVED** [review]
I've checked this PR for the second time and I'll merge this today. It would be great if somebody could take another look.

**Pull request merged**